daffodil-core/src/test/scala/org/apache/daffodil/api/TestParseIndividualMessages.scala

```diff
@@ -0,0 +1,230 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.api
+
+import org.apache.daffodil.Implicits.intercept
+import org.apache.daffodil.io.SocketPairTestRig
+import org.apache.daffodil.util.SchemaUtils
+import org.apache.daffodil.util.TestUtils
+import org.junit.Assert._
+import org.junit.Test
+
+import java.io.InputStream
+import java.io.OutputStream
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.concurrent.TimeoutException
+import scala.concurrent.duration.DurationInt
+import scala.xml.Node
+
+
+/**
+ * Shows that we can parse exactly 1 message from a TCP network socket
+ * without blocking for bytes past the end of the messsage.
+ *
+ * This only works for DFDL schemas of formats that are specified length.
+ */
+class TestParseIndividualMessages {
+
+  import SocketPairTestRig._
+
+  //
+  // DFDL schema for element e1 which occupies exactly 4 bytes.
+  //
+  val exactly4ByteSch = SchemaUtils.dfdlTestSchema(
+      <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd"/>,
+      <dfdl:format representation="binary" byteOrder="bigEndian" binaryNumberRep="binary" ref="tns:GeneralFormat"/>,
+      <xs:element name="e1" type="xs:string" dfdl:lengthKind="explicit" dfdl:length="4"/>)
+
+  /**
+   * Test shows that at least for simple fixed-length data, Daffodil parse returns
+   * without requiring more bytes to be read than the exact length required.
+   */
+  @Test def testDaffodilParseFromNetwork1(): Unit = {
+    val sptr = new SocketPairTestRig {
+      override def test(pos: OutputStream, cis: InputStream): Unit = {
+
+        val dp = TestUtils.compileSchema(exactly4ByteSch)
+
+        //
+        // Write exactly 4 bytes to producer network stream
+        //
+        pos.write("1234".getBytes)
+        pos.flush()
+
+        //
+        // Daffodil parse element e1 from consumer input stream
+        //
+        // If we need more than 4 bytes to successfully parse (we shouldn't for this schema)
+        // then this will hang, because only 4 bytes are in fact available.
+        //
+        // Caution: if debugging, this will timeout if you stop inside here!
+        //
+        val (pr: DFDL.ParseResult, xml: Node) =
+        withTimeout("Daffodil parse") {
+          TestUtils.runDataProcessorOnInputStream(dp, cis, areTracing = false)
+        }
+
+        assertFalse(pr.isError)
+        assertEquals("1234", xml.text)
+      }
+    }
+    sptr.run()
+  }
+
+  /**
+   * Test shows that the isAtEnd method does not block looking for additional data.
```

Comment:
Comment now wrong. The underlying code isn't calling isAtEnd. This shows that reading 4 bytes does not block looking for additional data. 

This test is identical to the test above now, save that last assert. Combine these two tests.